Stackoverflow and too many open files

**Elasticsearch version** 6.8.6 docker

**Plugins installed**: []

**JVM version** (`java -version`):

**OS version** ubuntu 16

**Description of the problem including expected versus actual behavior**:

a cluster with 6 data nodes, 3 master nodes and 2 coordinating - total of 7.5K shards
each data node with 30GB of RAM, and max open files 65536

getting Too many open files errors and then stackoverflow exception.
not sure why getting the Too many open files exception...

seeing many nested exceptions like:

`
[2020-04-28T12:34:49,649][WARN ][o.e.i.e.Engine           ] [fmyinf7113-1] [mars_storage-directories_history_4b895ae4-c01d-4904-88b9-d11090dea4c6-000001][0] giving up looking for fatal errors
java.nio.file.FileSystemException: /usr/share/elasticsearch/data/nodes/0/indices/2E5BQgAiQ_u_EBSxl-JZlA/0/translog/translog-2167182483755916320.ckp: Too many open files
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:100) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116) ~[?:?]
        at sun.nio.fs.UnixCopyFile.copyFile(UnixCopyFile.java:248) ~[?:?]
        at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:603) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258) ~[?:?]
        at java.nio.file.Files.copy(Files.java:1298) ~[?:?]
        at org.elasticsearch.index.translog.Translog.copyCheckpointTo(Translog.java:285) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.index.translog.Translog.rollGeneration(Translog.java:1709) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.index.engine.InternalEngine.flush(InternalEngine.java:1818) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.index.shard.IndexShard.flush(IndexShard.java:1117) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.indices.flush.SyncedFlushService.performPreSyncedFlush(SyncedFlushService.java:507) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.indices.flush.SyncedFlushService.access$500(SyncedFlushService.java:74) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.indices.flush.SyncedFlushService$PreSyncedFlushTransportHandler.messageReceived(SyncedFlushService.java:817) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.indices.flush.SyncedFlushService$PreSyncedFlushTransportHandler.messageReceived(SyncedFlushService.java:813) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:30) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) [elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:692) [elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751) [elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.8.6.jar:6.8.6]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
        Suppressed: org.apache.lucene.store.AlreadyClosedException: translog [39] is already closed (path [/usr/share/elasticsearch/data/nodes/0/indices/2E5BQgAiQ_u_EBSxl-JZlA/0/translog/translog-39.tlog]
                at org.elasticsearch.index.translog.TranslogWriter.closeIntoReader(TranslogWriter.java:305) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.index.translog.Translog.rollGeneration(Translog.java:1706) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.index.engine.InternalEngine.flush(InternalEngine.java:1818) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.index.shard.IndexShard.flush(IndexShard.java:1117) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.indices.flush.SyncedFlushService.performPreSyncedFlush(SyncedFlushService.java:507) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.indices.flush.SyncedFlushService.access$500(SyncedFlushService.java:74) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.indices.flush.SyncedFlushService$PreSyncedFlushTransportHandler.messageReceived(SyncedFlushService.java:817) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.indices.flush.SyncedFlushService$PreSyncedFlushTransportHandler.messageReceived(SyncedFlushService.java:813) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:30) ~[elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) [elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:692) [elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751) [elasticsearch-6.8.6.jar:6.8.6]
                at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.8.6.jar:6.8.6]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
                at java.lang.Thread.run(Thread.java:830) [?:?]
        Caused by: java.nio.file.FileSystemException: /usr/share/elasticsearch/data/nodes/0/indices/2E5BQgAiQ_u_EBSxl-JZlA/0/translog/translog-2167182483755916320.ckp: Too many open files
                at sun.nio.fs.UnixException.translateToIOException(UnixException.java:100) ~[?:?]
                at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
                at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116) ~[?:?]
                at sun.nio.fs.UnixCopyFile.copyFile(UnixCopyFile.java:248) ~[?:?]
                at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:603) ~[?:?]
`

and in the end 
`
 at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:987) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.common.io.stream.StreamOutput.writeException(StreamOutput.java:985) ~[elasticsearch-6.8.6.jar:6.8.6]
        at org.elasticsearch.ElasticsearchException.writeStackTraces(ElasticsearchException.java:743) ~[elasticsearch-6.8.6.jar:6.8.6]

`
**Provide logs (if relevant)**:

--- a/server/src/main/java/org/elasticsearch/ExceptionsHelper.java
+++ b/server/src/main/java/org/elasticsearch/ExceptionsHelper.java
@@ -210,7 +210,7 @@ public final class ExceptionsHelper {
     }
 
     @SuppressWarnings("unchecked")
-    private static <T extends Throwable> Optional<T> unwrapCausesAndSuppressed(Throwable cause, Predicate<Throwable> predicate) {
+    public static <T extends Throwable> Optional<T> unwrapCausesAndSuppressed(Throwable cause, Predicate<Throwable> predicate) {
         if (predicate.test(cause)) {
             return Optional.of((T) cause);
         }

--- a/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
+++ b/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
@@ -1680,6 +1680,7 @@ public class InternalEngine extends Engine {
                         refresh("version_table_flush", SearcherScope.INTERNAL, true);
                         translog.trimUnreferencedReaders();
                     } catch (AlreadyClosedException e) {
+                        failOnTragicEvent(e);
                         throw e;
                     } catch (Exception e) {
                         throw new FlushFailedEngineException(shardId, e);

--- a/server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java
+++ b/server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java
@@ -37,10 +37,11 @@ public class TragicExceptionHolder {
             return; // first exception
         }
         final Exception tragedy = this.tragedy.get();
-        assert ex == tragedy || ex instanceof AlreadyClosedException : new AssertionError("must be ACE or tragic exception", ex);
         // ensure no circular reference
         if (ExceptionsHelper.unwrapCausesAndSuppressed(ex, e -> e == tragedy).isEmpty()) {
             tragedy.addSuppressed(ex);
+        } else {
+            assert ex == tragedy || ex instanceof AlreadyClosedException : new AssertionError("must be ACE or tragic exception", ex);
         }
     }
 

--- a/server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java
+++ b/server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java
@@ -19,6 +19,9 @@
 
 package org.elasticsearch.index.translog;
 
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.ExceptionsHelper;
+
 import java.util.concurrent.atomic.AtomicReference;
 
 public class TragicExceptionHolder {
@@ -30,10 +33,14 @@ public class TragicExceptionHolder {
      */
     public void setTragicException(Exception ex) {
         assert ex != null;
-        if (tragedy.compareAndSet(null, ex) == false) {
-            if (tragedy.get() != ex) { // to ensure there is no self-suppression
-                tragedy.get().addSuppressed(ex);
-            }
+        if (tragedy.compareAndSet(null, ex)) {
+            return; // first exception
+        }
+        final Exception tragedy = this.tragedy.get();
+        assert ex == tragedy || ex instanceof AlreadyClosedException : new AssertionError("must be ACE or tragic exception", ex);
+        // ensure no circular reference
+        if (ExceptionsHelper.unwrapCausesAndSuppressed(ex, e -> e == tragedy).isEmpty()) {
+            tragedy.addSuppressed(ex);
         }
     }
 

--- a/server/src/main/java/org/elasticsearch/index/translog/Translog.java
+++ b/server/src/main/java/org/elasticsearch/index/translog/Translog.java
@@ -1623,10 +1623,9 @@ public class Translog extends AbstractIndexShardComponent implements IndexShardC
      * @throws IOException if an I/O exception occurred during any file operations
      */
     public void rollGeneration() throws IOException {
-        // make sure we move most of the data to disk outside of the writeLock
-        // in order to reduce the time the lock is held since it's blocking all threads
-        sync();
+        syncBeforeRollGeneration();
         try (Releasable ignored = writeLock.acquire()) {
+            ensureOpen();
             try {
                 final TranslogReader reader = current.closeIntoReader();
                 readers.add(reader);
@@ -1641,6 +1640,12 @@ public class Translog extends AbstractIndexShardComponent implements IndexShardC
                 throw e;
             }
         }
+    }
+
+    void syncBeforeRollGeneration() throws IOException {
+        // make sure we move most of the data to disk outside of the writeLock
+        // in order to reduce the time the lock is held since it's blocking all threads
+        sync();
     }
 
     /**

--- a/server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java
+++ b/server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java
@@ -100,9 +100,12 @@ import java.util.Comparator;
 import java.util.Deque;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.IdentityHashMap;
 import java.util.Iterator;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Queue;
 import java.util.Set;
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
@@ -3262,5 +3265,63 @@ public class TranslogTests extends ESTestCase {
                 thread.join();
             }
         }
+    }
+
+    public void testEnsureNoCircularException() throws Exception {
+        final AtomicBoolean failedToSyncCheckpoint = new AtomicBoolean();
+        final ChannelFactory channelFactory = (file, openOption) -> {
+            final FileChannel channel = FileChannel.open(file, openOption);
+            return new FilterFileChannel(channel) {
+                @Override
+                public void force(boolean metaData) throws IOException {
+                    if (failedToSyncCheckpoint.get()) {
+                        throw new IOException("simulated");
+                    }
+                    super.force(metaData);
+                }
+            };
+        };
+        final TranslogConfig config = getTranslogConfig(createTempDir());
+        final String translogUUID = Translog.createEmptyTranslog(
+            config.getTranslogPath(), SequenceNumbers.NO_OPS_PERFORMED, shardId, channelFactory, primaryTerm.get());
+        final Translog translog = new Translog(config, translogUUID, new TranslogDeletionPolicy(),
+            () -> SequenceNumbers.NO_OPS_PERFORMED, primaryTerm::get,
+            seqNo -> {}) {
+            @Override
+            ChannelFactory getChannelFactory() {
+                return channelFactory;
+            }
+
+            @Override
+            void syncBeforeRollGeneration() {
+                // make it a noop like the old versions
+            }
+        };
+        try (translog) {
+            translog.add(new Translog.Index("1", 1, primaryTerm.get(), new byte[]{1}));
+            failedToSyncCheckpoint.set(true);
+            expectThrows(IOException.class, translog::rollGeneration);
+            final AlreadyClosedException alreadyClosedException = expectThrows(AlreadyClosedException.class, translog::rollGeneration);
+            if (hasCircularReference(alreadyClosedException)) {
+                throw new AssertionError("detect circular reference exception", alreadyClosedException);
+            }
+        }
+    }
+
+    static boolean hasCircularReference(Exception cause) {
+        final Queue<Throwable> queue = new LinkedList<>();
+        queue.add(cause);
+        final Set<Throwable> seen = Collections.newSetFromMap(new IdentityHashMap<>());
+        while (queue.isEmpty() == false) {
+            final Throwable current = queue.remove();
+            if (seen.add(current) == false) {
+                return true;
+            }
+            Collections.addAll(queue, current.getSuppressed());
+            if (current.getCause() != null) {
+                queue.add(current.getCause());
+            }
+        }
+        return false;
     }
 }

